server.c

```diff
@@ -3570,6 +3603,88 @@ port_is_zero(
 #endif
 }
 
+/* Parses the PROXYv2 header from buf and updates the struct.
+ * Returns 1 on success, 0 on failure. */
+static int
+consume_pp2_header(struct buffer* buf, struct query* q, int stream)
+{
+	size_t size;
+	struct pp2_header* header;
+	int err = pp2_read_header(buffer_begin(buf), buffer_remaining(buf));
+	if(err) {
+		VERBOSITY(4, (LOG_ERR, "proxy-protocol: could not parse "
+			"PROXYv2 header: %s", pp_lookup_error(err)));
+		return 0;
+	}
+	header = (struct pp2_header*)buffer_begin(buf);
+	size = PP2_HEADER_SIZE + read_uint16(&header->len);
+	if(size > buffer_limit(buf)) {
+		VERBOSITY(4, (LOG_ERR, "proxy-protocol: not enough buffer "
+			"size to read PROXYv2 header"));
+		return 0;
+	}
+	if((header->ver_cmd & 0xF) == PP2_CMD_LOCAL) {
+		/* A connection from the proxy itself.
+		 * No need to do anything with addresses. */
+		goto done;
+	}
+	if(header->fam_prot == 0x00) {
```

Comment:
This is also in the Unbound proxy protocol code. I think it is because the 0x00 check is for an out of band message, of a healthcheck type, and it does not contain an address. In the switch it deals with the address, so this skips the address management part, to do other commands.

Apart from that, the code could be in the switch, technically. So I would not mind really, either way. I'll keep it like it is now, because of the flow of the algorithm from the comments.